File: Dart/202.happy_number.dart
void main() {
  print(isHappy(19));
  print(isHappy(2));
}

bool isHappy(int n) {
  final set = <int>{};

  while (n > 1) {
    if (!set.add(n)) {
      break;
    }

    final parStr = n.toString().split("");

    n = 0;
    for (int i = 0; i < parStr.length; i++) {
      final value = int.parse(parStr[i]);
      n += value * value;
    }
  }

  return n == 1;
}

File: src/202.happy_number.kt
fun main() {
    println(isHappy(19))
    println(isHappy(2))
}

fun isHappy(n: Int): Boolean {
    var x = n
    val set = mutableSetOf<Int>()

    while (x > 1) {
        if (!set.add(x)) {
            break
        }

        val str = x.toString()
        x = 0

        for (s in str) {
            x += "$s".toInt() * "$s".toInt()
        }
    }

    return x == 1
}